windows: Add windows build

yincrash opened the pull request:
This requires the builder to supply a Segoe UI Symbol input font
and will merge the regular font with the input font.

For #15

yincrash commented:
Hmm. Maybe I should change this to be a Windows batch script that can just take in the EmojiOne and Segue UI fonts? So a user wouldn't have to attempt to use the build system on Windows.

13rac1 commented on windows/generate_cmap4_from_12.py:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python2
```
What is the license on this code? Did you write all of it?

yincrash replied:
I used this [sample script](https://github.com/behdad/fonttools/blob/master/Snippets/cmap-format.py) in font tools as inspiration. It's under an MIT style license as well, however I'm not sure what the guideline is for attribution here.

13rac1 replied:
Behdad Esfahbod wrote it. So, I'd add a copy of the MIT license copyright him.

13rac1 commented on Makefile:
```diff
@@ -73,13 +77,36 @@ osx-package: $(OSX_FONT)
 	7z a -tzip -mx=9 $(OSX_PACKAGE).zip ./$(OSX_PACKAGE)
 
+windows-package: $(WINDOWS_FONT)
+	cp $(WINDOWS_FONT) .
```
Don't necessarily need a `windows-package` since there's no "package" it's not making a zip/tar.gz.

yincrash replied:
I considered not, but it needed a .PHONY target for easy building, and it made it consistent with the other packages where the final result ended up in the top level directory.

13rac1 commented on Makefile:
```diff
@@ -73,13 +77,36 @@ osx-package: $(OSX_FONT)
+$(WINDOWS_FONT): $(TMP)/svg.ttx $(TMP)/name.ttx $(REGULAR_FONT)
+	# merge segoe ui symbol into emoji one, this keeps emoji one's glyph ids intact
+	# for the 'SVG ' table data
+	pyftmerge $(REGULAR_FONT) $(SEGOE_UI_FONT)
```
Additional dependencies need to be added to the README.md

yincrash replied:
this is in the FontTools package including ttx.

13rac1 replied:
Hmm... Not on my machine. I have ttx, but not pyftmerge. Maybe it's not in the Ubuntu package. How did you install fonttools?

yincrash replied:
`pip install fonttools`

the [changelog](https://github.com/behdad/fonttools/blob/master/NEWS) says pyftmerge showed up in 2.5.

13rac1 replied:
HAHA! I had both the pip and the apt packages installed.

eosrei commented:
#1 Thanks for a great PR! Even followed my git commit message standard 😍

> Maybe I should change this to be a Windows batch script that can just take in the EmojiOne and Segue UI fonts?

Yes, I think so. The overall build doesn't run on Windows. Right now I am thinking it would be best as a batch file in a ` EmojiOneColor-SVGinOT-OSX-1.3.zip` file. People will be able to download the zip, then ideally just run the batch file (which can check the default source ttf location).

eosrei commented:
Could you rebase this into one commit and force push to this branch? It'll be something like:
```
git rebase -i b6d5340
# Set your original commit to squash, then save
git push -f
```

eosrei commented:
This will need install directions added to the README. What versions of Windows does this support? 8 and 10?

yincrash commented:
I meant to push this to a -wip branch so it wouldn't hit the PR yet. Will do on the squashing. I'm downloading Microsoft's testing VM to try this on 8 and 10, since I only use 7. I think I might have to edit it a bit because I think they use a different font name. I'll add to the README once I know that's settled.

yincrash commented:
Tested on both 8 and 10. On those versions, both the Emoji and the Symbol font need to be replaced.

Also updated the README.

13rac1 commented on README.md:
```diff
@@ -153,13 +153,24 @@ Chrome will use the fallback black and white emoji.*
-from: https://github.com/eosrei/emojione-color-font/releases
-
-Help wanted: [How to override the Windows default Segoe UI Emoji font][16]
+There are two install options for OS X. Both SVGinOT versions are available
```
>  OS X

 ;)

yincrash replied:
Whoops! Fixed.

eosrei commented:
I'll test this later this week. Thank you for your work on it! 😉

eosrei commented:
Just found this. You might have made a solution ;) http://superuser.com/questions/1062418/how-do-i-replace-windows-10s-emoji

13rac1 commented on windows/install.cmd:
```diff
@@ -0,0 +1,114 @@
+REM Merge Segoe UI Symbol into EmojiOne, this keeps emoji one's glyph ids intact
+REM for the 'SVG ' table data
+pyftmerge %EMOJI_FONT_PATH% %MS_FONT_PATH%
+REM pyftmerge doesn't generate a cmap4 table if either font has a cmap12 table
```
Is there any way we can fix this in SCFBuild?

yincrash replied:
I actually fixed this in [FontTools](https://github.com/behdad/fonttools/pull/635), so it will be in whatever their next release is. The emoji font has the correct tables, it was FontTool's fault that the merge was incorrect. That's what the python script corrects, and what my PR for them fixed.

13rac1 commented on windows/install.cmd:
```diff
@@ -0,0 +1,114 @@
+python %CMAP_SCRIPT% "merged.ttf" "mergedc4.ttf" || GOTO :ERROR
+DEL "merged.ttf"
+ECHO Dumping SVG emojis
+ttx -t "SVG " -o "svg.ttx" %EMOJI_FONT_PATH% || GOTO :ERROR
```
I know it's not _necessary_, but these should probably all go in `%Temp%`. Thoughts?

yincrash replied:
All of the work does happen in `%TEMP%`, that's what `PUSHD %TEMP%` and it's matching `POPD` does. `pyftmerge` only outputs files to the current working directory, so I change the working directory to %TEMP% for all these commands.

13rac1 replied:
Ah ha! Perfect! I'm not proficient in the new Windows scripting, I didn't even realize that.

eosrei commented:
`EmojiOneColor-SVGinOT-Win-1.3.zip` is built. I'll try it on my Win10 desktop tomorrow. Thank you :smile:

eosrei commented:
I'm using Windows for the first time this year. What have you done to me ?!?! Haha! 😲

I removed all the old installs of everything in Win10. Then:

1. Fresh install of Chocolately
2. `choco install python2`
3. `pip` ... Nope. `pip` isn't in the system path? O_o
4. Searching for pip...
5. `c:\tools\python2\scripts\pip install fonttools`
6. `python`
7. Test fonttools: `import fontTools` `print fonTools.version` "3.0" good.
8.  `cd .\EmojiOneColor-SVGinOT-Win-1.3\`
9. `.\install.cmd`
```
FontTools 3.x is not available in PATH. Is it installed?
https://github.com/behdad/fonttools
Installation failed!
```

Adding `C:\tools\python2\Scripts` to my PATH makes everything work. Nice script!

We need this to work "out of the box". Step by step for the non-experienced user. Can you add more details to the install process in the README.md? I'd recommend Chocolatey to get people using it.


*Side note: Powershell is so slow. Just typing `dir` and waiting for it to think about it. I'll stay with `cmd.exe` I guess? I'm actually quite surprised.*

yincrash commented:
Ahhh. I installed python using the installer from python.org. That's really unfortunate the Chocolatey package does not include the scripts folder in the path. Would you recommend just ensuring those directions in the README or try to handler some of these different cases in the script?

eosrei commented:
I'll uninstall and try it again with the python.org installer. Chocolatey just runs the installer, it shouldn't make a difference.

eosrei commented:
I found the cause of this problem. I setup a fresh win10 VM (https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) and installed the current Python 2.7 from python.org. Found this:
![Path not default](https://cloud.githubusercontent.com/assets/195061/16921964/94c12e1a-4cc8-11e6-8d62-7fbdde67acdc.png)
The path isn't automatically set and it's "below the fold" so the majority of inexperienced users aren't going to select it.

I've discovered Chocolatey will set the path correctly if the `easy.install` package is installed.
```
choco install python2
choco install easy.install
```

I'm trying to figure out a long term straightforward solution for this. I've looked [PyInstaller](http://www.pyinstaller.org/), but that seems overkill. I tested installing Fonttools from git:
```
pip install git+git://github.com:behdad/fonttools.git
```
It gets the current repo version which has [your fixes](https://github.com/behdad/fonttools/pull/635) We could avoid packaging the `generate_cmap4_from_12.py` script this way.

This is all just notes. Nothing actionable.

**Solution?** Best thing right now is a step-by-step install process. The Linux manual install process, for example, includes everything needed. It's just a copy/paste. We need something similar for Windows. 

Thanks for your work on this! I need this entire process tested, documented and polished since I need to support this going forward. Thanks for understanding :smile:

yincrash commented:
I totally understand. Time spent getting it right now, means less time dealing with users who don't get it right when trying to install.

What about just having the script install the FontTools from git source and ensure the scripts directory is in the path?

eosrei commented:
> What about just having the script install the FontTools from git source

Yes. Sounds easier to maintain. It's possible to avoid a dependency on git.:
```
pip install https://github.com/behdad/fonttools/archive/master.zip
```

> and ensure the scripts directory is in the path?

Yes. It could be temporarily added.

Rough draft of instructions:
1. Download Python 3 for windows: https://www.python.org/downloads/
2. Install the MSI (include path add details?)
3. Download EmojiOne Color font from releases: https://github.com/eosrei/emojione-color-font/releases
4. Unzip the file
5. Open an Administrator CMD.exe
6. cd to uncompressed EmojiOneColor directory
7. `pip install https://github.com/behdad/fonttools/archive/master.zip`
8. `install.cmd` This will take some time. Install the new fonts when requested.
9. Done!

Still so complex...

yincrash commented:
I don't think it's necessary to run an Administrator command prompt. The font viewer should have a little admin permission request box when attempting to install fonts.

eosrei commented:
Oh good point. I was thinking it was needed for pip, but no.

yincrash commented:
Okay, I'll edit the script to temporarily set the PATH variable, and run the `pip install` as well. That will remove step 7.

yincrash commented:
I can tackle this tomorrow. Things have been busy, and I want to test and make sure it will work the correctly. :)

yincrash commented:
Ugh. So, unfortunately, a foolproof way of determining where python is installed if it is not on the PATH is not easy, but it is completely outlined in [PEP 514](https://www.python.org/dev/peps/pep-0514/). There is a small [app](https://bitbucket.org/vinay.sajip/pylauncher) that we could bundle (created by python devs and BSD 2-clause licensed) to do this work for us of determining the install location. However, I think this is a bit overkill, but doing the search in the install script will also be very difficult, and not pretty looking at all.

It might just make more sense to check if python is available on the PATH, and if it isn't, instruct the user to add it to the PATH (as a message in the install script when it exits on error).

Suggestions?

eosrei commented:
In Windows 10: I've tried installing Python2 via Chocolately and the installer, then Python3 via Chocolately and the the installer. Chocolately with Python3 gets close to useful. `python` and `pip` work out of the box, but Chocolatey adds links to the specific executables only. Installing `fonttools` results in a non-accessible `ttx`.

The best option is Python *3* with the [regular installer](https://www.python.org/downloads/windows/). The user needs to specifically check the install to path, then `python` and `pip` are added to the path:
![Add python/pip to path](https://cloud.githubusercontent.com/assets/195061/17085663/c8a66f46-5192-11e6-8205-eab47ee9db47.png)

Having Python and Pip in the path should be a hard requirement.
* `Python.exe not found, install or add to PATH.`
* `Pip.exe not found, install or add to PATH.`

The instructions would say to install the most recent Python3, then `install.cmd` can just fail with the above errors if not found.

eosrei commented:
Second *draft* of instructions:

1. Download the most recent Python 3 for Windows: https://www.python.org/downloads/windows/ (It would be neat if we could figure out their OS detection URL and just link that)
2. Start the installer, select "Add Python 3.5 to PATH" and finish the install process.
3. Download EmojiOne Color font from releases: https://github.com/eosrei/emojione-color-font/releases
4. Uncompress the file
6. Open the new EmojiOneColor directory
7. Run `install.cmd`. Note: This will take some time.
8. Install the new fonts when requested.
9. Done!

yincrash commented:
Install script now checks for python and pip, then installs the master.zip of FontTools. Removed the cmap script now that we no longer need it. Updated the README

13rac1 commented on windows/install.cmd:
```diff
@@ -0,0 +1,115 @@
+
+ECHO Ensuring the latest FontTools is installed.
+
+pip install --upgrade https://github.com/behdad/fonttools/archive/master.zip
```
Can you add a call to `WHERE` for `ttx`? Might as well just check that after the `fonttools` install, confirms everything is working.

Otherwise this is looking really good! Thank you

yincrash replied:
Done

13rac1 replied:
Nice! I'll rebuild everything and check it tonight on Win10

eosrei commented:
I've tested install and uninstall in Windows 10. Everything seems to be working, even the new power symbols! ⏻ ⏼ ⏽ ⭘ ⏾

I'm happy to merge this as-is now. :smile: Great work!

Separate/additional questions:

1. Would it be possible to automatically install the fonts using this method? http://windowsitpro.com/scripting/trick-installing-fonts-vbscript-or-powershell-script

2. Could you make a PR against https://github.com/eosrei/twemoji-color-font for the same thing? Let me know, thank you.

yincrash commented:
Hmm. Yeah, that should work. Both the VBS and PS method should work fine. For eosrei/twemoji-color-font, can you make an issue and assign it to me? I can handle it, but it might be in a week or two (or this week if I'm lucky).

eosrei commented:
Cool!

I didn't want to wait any longer! https://github.com/eosrei/emojione-color-font/releases/tag/v1.3
